File: Retos/Reto #3 - EL GENERADOR DE CONTRASEÑAS [Media]/python/dPenedo.py
# /*
#  * Escribe un programa que sea capaz de generar contraseñas de forma aleatoria.
#  * Podrás configurar generar contraseñas con los siguientes parámetros:
#  * - Longitud: Entre 8 y 16.
#  * - Con o sin letras mayúsculas.
#  * - Con o sin números.
#  * - Con o sin símbolos.
#  * (Pudiendo combinar todos estos parámetros entre ellos)
#  */


import random
import string

print("*"*40)
print("Bienvenido al generador de contraseñas")
print("*"*40)


def respuesta_valida_sn(pregunta):
    print("Por favor, responde con 's' para Sí y 'n' para No.")
    while True:
        respuesta = input(pregunta).lower()
        if respuesta =="s":
            return True
        elif respuesta == "n":
            return False
        else:
            print("La respuesta debe ser 's' o 'n")

def generar_letra(tipo):
    letra_aleatoria = random.choice(tipo)
    lista_password.append(letra_aleatoria)


while True:
    print("Elige los siguientes parámetros")
    print("-"*40)
    longitud = input("¿Entre 8 y 16, qué longitud quieres que tenga la contraseña? Escribe un número del 8 al 16 \n")
    if longitud.isnumeric() == False:
        print("Debes introducir un número")
        continue
    elif int(longitud) not in range(8, 16):
        print("Debes introducir un número entre el 8 y el 16")
        continue

    mayus = respuesta_valida_sn("¿Quieres que tenga mayúsculas? ")
    numeros = respuesta_valida_sn("¿Quieres que tenga números? ")
    simbolos = respuesta_valida_sn("¿Quieres que tenga simbolos?")
    
    lista_password = []

    tipo_letras = string.ascii_letters if mayus else string.ascii_lowercase
    if numeros:
        tipo_letras += string.digits
    if simbolos:
        tipo_letras += string.punctuation

    for n in range(int(longitud)):
        generar_letra(tipo_letras)
    password = ''.join(lista_password)
    print("-"*40)
    print(f"Tu contraseña generada es: {password} ")
    print(f"longitud = {longitud} | mayus = {mayus} | numeros = {numeros} | simbolos = {simbolos}")
    print("-"*40)
    break

File: Retos/Reto #3 - EL GENERADOR DE CONTRASEÑAS [Media]/javascript/dpenedo.js
// Escribe un programa que sea capaz de generar contraseñas de forma aleatoria.
// Podrás configurar generar contraseñas con los siguientes parámetros:
// - Longitud: Entre 8 y 16.
// - Con o sin letras mayúsculas.
// - Con o sin números.
// - Con o sin símbolos.
// (Pudiendo combinar todos estos parámetros entre ellos)

// Configuraciones:
const passwordLength = 8;
const withUpperCase = true;
const withNumbers = true;
const withSymbols = true;

function generateTextDescriptions() {
    let textUpperCase;
    let textNumbers;
    let textSymbols;
    if (withUpperCase) {
        textUpperCase = 'has uppercase characters';
    } else {
        textUpperCase = 'has not uppercase characters';
    }
    if (withNumbers) {
        textNumbers = 'has numbers';
    } else {
        textNumbers = 'has not numbers';
    }
    if (withSymbols) {
        textSymbols = 'has symbol characters';
    } else {
        textSymbols = 'has not symbol characters';
    }
    return [textUpperCase, textNumbers, textSymbols];
}

function addRandomCharacter() {
    const lowercaseChars = 'abcdefghijklmnopqrstuvwxyz';
    const uppercaseChars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    const numberChars = '0123456789';
    const symbolChars = '!@#$%^&*()-_=+';

    let chars = lowercaseChars;
    if (withUpperCase) {
        chars += uppercaseChars;
    }
    if (withNumbers) {
        chars += numberChars;
    }
    if (withSymbols) {
        chars += symbolChars;
    }
    const randomCharachter = chars[Math.floor(Math.random() * chars.length)];

    return randomCharachter;
}

function generatePassword() {
    const texts = generateTextDescriptions();
    const [textUpperCase, textNumbers, textSymbols] = texts;
    let password = '';
    console.log('Welcome to the password generator');
    console.log(
        `Your password has ${passwordLength} elements ${textUpperCase}, ${textNumbers} and ${textSymbols}`,
    );
    for (let i = 0; i < passwordLength; i++) {
        password += addRandomCharacter();
    }
    console.log(`*********************************
    Your password is: ${password}
*********************************`);
}

generatePassword();
